src/main/java/sparksoniq/spark/iterator/flowr/LetClauseSparkIterator.java

```diff
@@ -36,25 +38,113 @@
 import java.util.List;
 
 public class LetClauseSparkIterator extends FlowrClauseSparkIterator {
+
+    private VariableReferenceIterator _variableReference;
+    private String _variableName;           // for efficient use in local iteration
+    private RuntimeIterator _expression;
+    private DynamicContext _tupleContext;   // re-use same DynamicContext object for efficiency
+    private FlworTuple _nextLocalTupleResult;
+
     public LetClauseSparkIterator(RuntimeTupleIterator child, VariableReferenceIterator variableReference, RuntimeIterator expression, IteratorMetadata iteratorMetadata) {
         super(child, null, FLWOR_CLAUSES.LET, iteratorMetadata);
         this._children.add(variableReference);
         this._children.add(expression);
+        _variableReference = variableReference;
+        _expression = expression;
     }
-    
+
+    @Override
+    public boolean isRDD() {
+        if (this._child == null) {
+            return false;
+        } else {
+            return _child.isRDD();
+        }
+    }
+
+    @Override
+    public FlworTuple next() {
+        if(_hasNext == true){
+            FlworTuple result = _nextLocalTupleResult;      // save the result to be returned
+            setNextLocalTupleResult();              // calculate and store the next result
+            return result;
+        }
+        throw new IteratorFlowException("Invalid next() call in let flwor clause", getMetadata());
+    }
+
+    private void setNextLocalTupleResult() {
+        // if first let clause, there are no more tuples
+        if (this._child == null) {
+            this._hasNext = false;
+            return;
+        }
+
+        if (_child.hasNext()) {
+            FlworTuple inputTuple = _child.next();
+            _tupleContext.removeAllVariables();             // clear the previous variables
+            _tupleContext.setBindingsFromTuple(inputTuple);      // assign new variables from new tuple
+
+            List<Item> result = new ArrayList<>();
+            _expression.open(_tupleContext);
+            while (_expression.hasNext())
+                result.add(_expression.next());
+            _expression.close();
+
+            FlworTuple resultTuple = new FlworTuple();
+            resultTuple.putValue(_variableName, result, true);
+            _nextLocalTupleResult = resultTuple;
+            this._hasNext = true;
+        } else {
+            _child.close();
+            this._hasNext = false;
+        }
+    }
+
+    @Override
+    public void open(DynamicContext context) {
+        super.open(context);
+
+        // evaluate locally
+        if (!isRDD()) {
+            if (this._child != null) { //if it's not a start clause
+                _tupleContext = new DynamicContext(_currentDynamicContext);     // assign current context as parent
+                _child.open(_tupleContext);
```

Comment:
For consistency (even if it leads to the same result), I suggest

_child.open(_currentDynamicContext).

_tupleContext is meant to be the dynamic context that you update in place, for use on the child expression.